lib/membrane_rtmp_plugin/rtmp/sink/sink.ex

```diff
@@ -0,0 +1,197 @@
+                This URL should be provided by your streaming service."
+              ]
+
+  @impl true
+  def handle_init(options) do
+    if String.starts_with?(options.rtmp_url, @supported_protocols) do
+      {:ok, Map.from_struct(options) |> Map.merge(@default_state_attributes)}
+    else
+      raise("Invalid destination URL provided")
+    end
+  end
+
+  @impl true
+  def handle_prepared_to_playing(ctx, state) do
+    case Native.create(state.rtmp_url) do
+      {:ok, native} ->
+        Membrane.Logger.debug("Correctly initialized connection with: #{state.rtmp_url}")
+        state = Map.put(state, :native, native)
+        demands = ctx.pads |> Map.keys() |> Enum.map(&{:demand, &1})
+        {{:ok, demands}, state}
+
+      {:error, reason} ->
+        raise("Transition to state playing failed with reason: #{reason}")
+    end
+  end
+
+  @impl true
+  def handle_playing_to_prepared(_ctx, state) do
+    Native.finalize_stream(state.native)
+    state = Map.merge(state, @default_state_attributes)
+    Membrane.Logger.debug("Stream correctly closed")
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_caps(
+        :video,
+        %MP4.Payload{content: %MP4.Payload.AVC1{avcc: avc_config}} = caps,
+        _ctx,
+        state
+      ) do
+    case Native.init_video_stream(state.native, caps.width, caps.height, avc_config) do
+      {:ok, ready, native} ->
+        Membrane.Logger.debug("Correctly initialized video stream.")
+        state = Map.merge(state, %{native: native, ready: ready})
+        {:ok, state}
+
+      {:error, :caps_resent} ->
+        Membrane.Logger.error(
+          "Input caps redefined on pad :video. RTMP Sink does not support changing stream parameters"
+        )
+
+        {:ok, state}
+    end
+  end
+
+  @impl true
+  def handle_caps(:audio, %Membrane.AAC{} = caps, _ctx, state) do
+    profile = AAC.profile_to_aot_id(caps.profile)
+    sr_index = AAC.sample_rate_to_sampling_frequency_id(caps.sample_rate)
+    channel_configuration = AAC.channels_to_channel_config_id(caps.channels)
+    frame_length_id = AAC.samples_per_frame_to_frame_length_id(caps.samples_per_frame)
+
+    aac_config =
+      <<profile::5, sr_index::4, channel_configuration::4, frame_length_id::1, 0::1, 0::1>>
+
+    case Native.init_audio_stream(state.native, caps.channels, caps.sample_rate, aac_config) do
+      {:ok, ready, native} ->
+        Membrane.Logger.debug("Correctly initialized audio stream.")
+        state = Map.merge(state, %{native: native, ready: ready})
+        {:ok, state}
+
+      {:error, :caps_resent} ->
+        Membrane.Logger.error(
+          "Input caps redefined on pas :audio. RTMP Sink does not support changing stream paremeters"
+        )
+
+        {:ok, state}
+    end
+  end
+
+  @impl true
+  def handle_write(pad, buffer, _ctx, %{ready: false} = state) do
+    state = Map.put(state, :buffered_frame, {pad, buffer})
+    {:ok, state}
+  end
```

Comment:
why is this needed?

Reply:
Together with @daniel-jodlos and @bblaszkow06 we agreed that we must demand from both audio and video pads in handle_prepared_to_playing to avoid "magic" in membrane. If I demand from both pads the following scenario can occur (and indeed occurs quite frequently): sink might receive first video frame before receiving audio caps. Unfortunately I cannot send any frame before I receive both caps. Bot caps are needed to send correct stream header. Therefore we agreed with @daniel-jodlos and @bblaszkow06 that any "falstart" frame mus be buffered and send ASAP once both caps are received and stream header is correctly sent.

Reply:
So you can just wait until you get both caps and then make demands ;)

Reply:
Before the first round of code review it was implemented this way to wait for both caps before any demands. @daniel-jodlos insisted that we demand from handle_prepared_to_playing. I will consult with him it once again.

Reply:
cc @bblaszkow06 @daniel-jodlos

Reply:
> So you can just wait until you get both caps and then make demands

The problem is that this works only thanks to input buffer making an initial demand - there's no guarantee given by the framework that caps will arrive without making a demand (even though that's what will happen) and I believe it's a kind of magic that we want to avoid in Membrane

Reply:
The problem is that caps themselves are not being demanded, so from the sink perspective it's counter-intuitive that demand is required to receive them. I believe it should be the responsibility of each element to send caps and make a demand if it's needed to generate them. This is what we'd have to standardize, but it's more intuitive, easier to implement and potentially may speed up things, as caps could be generated before the sink requests any demands (even if no initial demand was generated)